localization/ndt_scan_matcher/src/map_update_module.cpp

```diff
@@ -107,6 +107,20 @@ bool MapUpdateModule::should_update_map(
   return distance > param_.update_distance;
 }
 
+bool MapUpdateModule::out_of_map_range(const geometry_msgs::msg::Point & position)
+{
+  if (last_update_position_ == std::nullopt) {
+    return true;
+  }
+
+  const double dx = position.x - last_update_position_.value().x;
+  const double dy = position.y - last_update_position_.value().y;
```

Comment:
The `last_update_position_` and `param_` in MapUpdateModule are accessed from both the callback_sensor_points and the callback_timer, making them susceptible to data races. 

Although it's not a critical issue at the moment, it poses a risk. Please add a lock or comments explaining why data races won't occur, and caution future developers who might work on this code.

Reply:
Reading an optional type can sometimes result in a bad optional access, so we need to be cautious. 
Here is an example.
```cpp
#include <iostream>
#include <optional>
#include <thread>
std::optional<int> count = std::nullopt;
void ThreadA()
{
  for (int i = 0; i < 100000; ++i) {
    if (i % 2 == 0)
      count = i;
    else
      count = std::nullopt;
  }
}
void ThreadB()
{
  long sum = 0;
  for (int i = 0; i < 100000; ++i) {
    if (count.has_value())
      sum += count.value(); // This will occur bad optinal access
  }
}
int main()
{
  std::thread th_a(ThreadA);
  std::thread th_b(ThreadB);
  th_a.join();
  th_b.join();
  return 0;
}

```

Reply:
@anhnv3991 
I think this is important, so please respond.

Reply:
mutex has already been added. So I close this thread.